cpp/src/strings/convert/convert_urls.cu

```diff
@@ -185,137 +192,184 @@ constexpr bool is_hex_digit(char ch)
   return (ch >= '0' && ch <= '9') || (ch >= 'A' && ch <= 'F') || (ch >= 'a' && ch <= 'f');
 }
 
-/**
- * @brief Functor for detecting character escape sequences in URL-encoded strings.
- */
-struct url_decode_escape_detector {
-  device_span<char const> const d_chars{};
-
-  /**
-   * @brief Detects if an escape sequence is at the specified character position.
-   *
-   * Returns true for a character index corresponding to the start of an escape
-   * sequence, i.e.: '%' followed by two hexadecimal digits. This does not check
-   * against string boundaries and therefore can produce false-positives.
-   *
-   * @param char_idx Character position to examine
-   * @return true An escape sequence was detected at the character position
-   * @return false No escape sequence at the character position
-   */
-  __device__ bool operator()(size_type char_idx) const
-  {
-    return (char_idx + 2 < d_chars.size()) && d_chars[char_idx] == '%' &&
-           is_hex_digit(d_chars[char_idx + 1]) && is_hex_digit(d_chars[char_idx + 2]);
-  }
-};
+__forceinline__ __device__ bool is_escape_char(char const* const ptr)
+{
+  return (ptr[0] == '%' && is_hex_digit(ptr[1]) && is_hex_digit(ptr[2]));
+}
 
 /**
- * @brief Functor for detecting escape sequence positions that cross a string boundary.
+ * @brief Count the number of characters of each string after URL decoding.
+ *
+ * @tparam threadblock_size Number of threads in a threadblock. This template argument must be a
+ * multiple of `cudf::detail::warp_size`.
```

Comment:
If you want to ensure complete warps, why not instead template on `num_warps`?

Reply:
Thanks for the review Vyas! Addressed in https://github.com/rapidsai/cudf/pull/8622/commits/961ed39a28b0bcac1f2d4abcc2b6855309535843.